pkg/apis/serving/v1alpha1/revision_validation.go

```diff
@@ -184,120 +190,193 @@ func ValidateContainerConcurrency(cc RevisionContainerConcurrencyType, cm Revisi
 		errs = errs.Also(apis.ErrMissingOneOf("secret", "configMap"))
 	}
 	return errs
 }
 
+func validateEnvValueFrom(source *corev1.EnvVarSource) *apis.FieldError {
+	if source == nil {
+		return nil
+	}
+	return validateDisallowedFields(*source, *EnvVarSourceMask(source))
+}
+
+func validateEnv(envVars []corev1.EnvVar) *apis.FieldError {
+	var errs *apis.FieldError
+	for i, env := range envVars {
+		errs = errs.Also(validateDisallowedFields(env, *EnvVarMask(&env)).ViaIndex(i)).Also(
+			validateEnvValueFrom(env.ValueFrom).ViaField("valueFrom").ViaIndex(i))
+	}
+	return errs
+}
+
+func validateEnvFrom(envFromList []corev1.EnvFromSource) *apis.FieldError {
+	var errs *apis.FieldError
+	for i, envFrom := range envFromList {
+		errs = errs.Also(validateDisallowedFields(envFrom, *EnvFromSourceMask(&envFrom)).ViaIndex(i))
+
+		cm := envFrom.ConfigMapRef
+		sm := envFrom.SecretRef
+		if sm != nil {
+			errs = errs.Also(validateDisallowedFields(*sm, *SecretMapEnvSourceMask(sm)).ViaIndex(i))
+			errs = errs.Also(validateDisallowedFields(
+				sm.LocalObjectReference, *LocalObjectReferenceMask(&sm.LocalObjectReference))).ViaIndex(i).ViaField("secretRef")
+		}
+
+		if cm != nil {
+			errs = errs.Also(validateDisallowedFields(*cm, *ConfigMapEnvSourceMask(cm)).ViaIndex(i))
+			errs = errs.Also(validateDisallowedFields(
+				cm.LocalObjectReference, *LocalObjectReferenceMask(&cm.LocalObjectReference))).ViaIndex(i).ViaField("configMapRef")
+		}
+		if cm != nil && sm != nil {
+			errs = errs.Also(apis.ErrMultipleOneOf("configMapRef", "secretRef"))
+		} else if cm == nil && sm == nil {
+			errs = errs.Also(apis.ErrMissingOneOf("configMapRef", "secretRef"))
+		}
+	}
+	return errs
+}
+
 func validateContainer(container corev1.Container, volumes sets.String) *apis.FieldError {
 	if equality.Semantic.DeepEqual(container, corev1.Container{}) {
 		return apis.ErrMissingField(apis.CurrentField)
 	}
 
+	errs := validateDisallowedFields(container, *ContainerMask(&container))
+
+	// Env
+	errs = errs.Also(validateEnv(container.Env).ViaField("env"))
+	// EnvFrom
+	errs = errs.Also(validateEnvFrom(container.EnvFrom).ViaField("envFrom"))
+	// Image
+	if _, err := name.ParseReference(container.Image, name.WeakValidation); err != nil {
+		fe := &apis.FieldError{
+			Message: "Failed to parse image reference",
+			Paths:   []string{"image"},
+			Details: fmt.Sprintf("image: %q, error: %v", container.Image, err),
+		}
+		errs = errs.Also(fe)
+	}
+	// Liveness Probes
+	errs = errs.Also(validateProbe(container.LivenessProbe).ViaField("livenessProbe"))
+	// Ports
+	errs = errs.Also(validateContainerPorts(container.Ports).ViaField("ports"))
+	// Readiness Probes
+	errs = errs.Also(validateProbe(container.ReadinessProbe).ViaField("readinessProbe"))
+	// Resources
+	errs = errs.Also(validateResources(&container.Resources).ViaField("resources"))
+	// SecurityContext
+	errs = errs.Also(validateSecurityContext(container.SecurityContext).ViaField("securityContext"))
```

Comment:
One thing that I don't like about this is that the fieldmask function and the validation function (here) have to stay in sync.